// adaptivejs/tasks/config/connect.js
module.exports = function(grunt){
    var enableCrossDomainRequests = function(connect, options, middlewares) {
            middlewares.unshift(function (req, res, next) {
                // CSOPS-103: Chrome now blocks cross origin requests for
                // resources
                // https://code.google.com/p/chromium/issues/detail?id=286681
                res.setHeader('Access-Control-Allow-Origin', '*');
                return next();
            });

            return middlewares;
        };

    return {
        http: {
            options: {
                hostname: '0.0.0.0',
                port: (grunt.option('port') || 8080),
                base: 'build/',
                middleware: enableCrossDomainRequests
            }
        },
        https: {
            options: {
                hostname: '0.0.0.0',
                protocol: 'https',
                port: (grunt.option('https-port') || 8443),
                base: 'build/',
                middleware: enableCrossDomainRequests
            }
        },
        test: {
            options: {
                hostname: '0.0.0.0',
                port: (grunt.option('p') || 8888),
                base: '.'
            }
        }
    };
};
