// grunt-mobify-cloud/tasks/verify-adaptive-dependencies.js
module.exports = function(grunt) {

    var fs = require('fs');
    var path = require('path');
    var Request = require('request');
    
    var Utils = require('../lib/utils');

    grunt.registerTask('mobify-verify_adaptive_dependencies',
            'Verifies dependencies against the Mobify Cloud', function() {
        var done = this.async();

        var options = this.options({
            projectSlug: '',
            basePath: process.cwd(),
            origin: 'https://cloud.mobify.com',
            settingsFile: Utils.getSettingsPath()
        });

        var uri = Utils.getProjectApiPath(
            options.origin, options.projectSlug,
            "adaptivejs/verify-dependencies/"
        );

        var form = {};
        ['package.json', 'bower.json'].forEach(function(filename) {
            form[filename] = fs.readFileSync(
                path.join(options.basePath, filename),
                { encoding: 'utf8' }
            );
        });

        var requestOptions = {
            auth: Utils.readCredentials(options.settingsFile),
            headers: Utils.getRequestHeaders(),
            form: form
        };

        Request.post(uri, requestOptions, function(err, response, body) {
            if (err) {
                throw err;
            }
            Utils.throwForStatus(response);

            var result = null;
            try {
                result = JSON.parse(body);
            } catch(e) {
                result = {
                    success: false,
                    message: "Could not verify adaptive dependencies, got response:\r\n" + body
                };
            }

            if (!result.success) {
                grunt.fail.warn(result.message);
            }

            done()
        });
    });
};
