// grunt-mobify-cloud/Gruntfile.js
/*
 * mobify-push
 * https://github.com/mobify/grunt-mobify-upload
 *
 * Copyright (c) 2013 Mobify
 */

'use strict';

var path = require('path');

var settingsPath = 'test/fixtures/creds';

module.exports = function(grunt) {

    // Project configuration.
    grunt.initConfig({
        jshint: {
            all: [
                'Gruntfile.js',
                'tasks/*.js',
                '<%= nodeunit.tests %>',
            ],
            options: {
                jshintrc: '.jshintrc',
            },
        },
        express: {
            custom: {
                options:{
                    port: 3000,
                    hostname: 'localhost',
                    server: 'test/mock-server.js'
                }
            }
        },
        'mobify-verify_adaptive_dependencies': {
            options: {
                projectSlug: 'test-project',
                basePath: path.join(process.cwd(), 'test/fixtures'),
                origin: 'https://localhost:3000',
                settingsFile: settingsPath
            }
        },
        'mobify-upload': {
            options: {
                origin: 'https://localhost:3000',
                projectSlug: 'test-project',
                settingsFile: settingsPath,
                message: grunt.option('message') || ''
            },
            src: ['test/**', '!test/fixtures/creds']
        },
        'mobify-save_credentials': {
            options: {
                settingsPath: './.mobify'
            }
        },
        nodeunit: {
            tests: ['test/test-*.js']
        },
        release: {
            options: {
                folder: '.',
                github: {
                    repo: 'mobify/grunt-mobify-cloud',
                    usernameVar: 'GITHUB_USERNAME',
                    passwordVar: 'GITHUB_TOKEN'
                }
            }
        }
    });

    grunt.loadTasks('tasks');

    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-contrib-nodeunit');
    grunt.loadNpmTasks('grunt-express');
    grunt.loadNpmTasks('grunt-release');

    grunt.registerTask('test', ['express', 'mobify-verify_adaptive_dependencies', 'mobify-upload', 'nodeunit']);
    grunt.registerTask('default', ['jshint', 'test']);
};
